components/servo/webview.rs

```diff
@@ -567,8 +577,107 @@ impl WebView {
         self.inner().javascript_evaluator.borrow_mut().evaluate(
             self.id(),
             script.to_string(),
+            None,
+            Box::new(callback),
+        );
+    }
+
+    /// Evaluate the specified string of JavaScript code. Once execution is complete or an error
+    /// occurs, Servo will call `callback`.
+    pub fn evaluate_javascript_with_message_port<T: ToString>(
+        &self,
+        script: T,
+        callback: impl FnOnce(Result<JSValue, JavaScriptEvaluationError>) + 'static,
+    ) -> Rc<MessagePort> {
+        let constellation_proxy = self.inner().constellation_proxy.clone();
+        let embedder_proxy = self.inner().embedder_proxy.clone();
+
+        let (port1, port2) = MessagePort::new_entangled(constellation_proxy, embedder_proxy);
+        self.inner().javascript_evaluator.borrow_mut().evaluate(
+            self.id(),
+            script.to_string(),
+            Some(port2),
             Box::new(callback),
         );
+        let port1 = Rc::new(port1);
+        self.inner_mut().message_ports.push(port1.clone());
+        port1
+    }
+
+    pub(crate) fn message_port(&self, message_port_id: MessagePortId) -> Option<Rc<MessagePort>> {
+        self.inner()
+            .message_ports
+            .iter()
+            .find(|port| port.id == message_port_id)
+            .cloned()
+    }
+}
+
+pub struct MessagePort {
+    pub(crate) id: MessagePortId,
+    pub(crate) entangled_id: MessagePortId,
+    //pub(crate) router_id: MessagePortRouterId,
+    constellation_proxy: ConstellationProxy,
+}
+
+impl MessagePort {
+    fn new_entangled(
+        constellation_proxy: ConstellationProxy,
+        embedder_proxy: EmbedderProxy,
+    ) -> (MessagePort, MessagePort) {
+        let (port_control_sender, port_control_receiver) =
+            ipc::channel().expect("ipc channel failure");
+        ROUTER.add_typed_route(
+            port_control_receiver,
+            Box::new(move |message| match message.unwrap() {
+                MessagePortMsg::CompleteTransfer(..) => unreachable!(),
+                MessagePortMsg::CompletePendingTransfer(..) => unreachable!(),
+                MessagePortMsg::CompleteDisentanglement(..) => unreachable!(),
+                MessagePortMsg::NewTask(message_port_id, task) => {
+                    let data = match task.data {
+                        PostMessageData::StructuredClone(..) => unreachable!(),
+                        PostMessageData::Serialized(data) => data,
+                    };
+                    let _ = embedder_proxy.send(EmbedderMsg::PostMessage(message_port_id, data));
+                },
```

Comment:
I think it's better to associate the `MessagePort` with a `WebViewId` and send it with `EmbedderMsg::PostMessage`, so that you do not have to iterate through all of the `WebView`s when handling this message.